File: misc/Equals.scala
import scala.util.control.Exception.catching

object Equals extends App {
  val s: String = null
  println(catching(classOf[NullPointerException]) opt {s equals "Fred"}) // Would boom! but catching returns None.
  println(catching(classOf[NullPointerException]) opt {"Fred" equals s})
  println(catching(classOf[NullPointerException]) opt {s == "Fred"})
  println(catching(classOf[NullPointerException]) opt {"Fred" == s})
}

File: misc/Equals.java
class Equals {
  static final void println(Object msg) { System.out.println(msg); }

  final boolean eqls(Object that) {
    System.out.println(this);
    return this.equals(that);
  }

  public static void main(String[] args) {
    Equals eN = null;
    Equals e1 = new Equals();
    println(eN);
    println(e1);

    println("e1 eqls eN");
    println(e1.eqls(eN));

    println("eN eqls e1");
    println(eN.eqls(e1)); // NullPointerException!
  }
}
